gcc/rust/lex/rust-input-source.h

```diff
@@ -0,0 +1,193 @@
+#ifndef RUST_INPUT_SOURCE_H
+#define RUST_INPUT_SOURCE_H
+
+#include "rust-codepoint.h"
+#include "optional.h"
+
+namespace Rust {
+// Input source wrapper thing.
+class InputSource
+{
+private:
+  // position of current character
+  unsigned int pos;
+  std::vector<Codepoint> chars;
+  bool is_valid_utf8;
+
+  // Overload operator () to return next char from input stream.
+  virtual int next_byte () = 0;
+
+  Codepoint next_codepoint ()
+  {
+    uint32_t input = next_byte ();
+
+    if ((int32_t) input == EOF)
+      return Codepoint::eof ();
+    else if (input < 128)
+      {
+	// ascii -- 1 byte
+	return {input};
+      }
+    else if ((input & 0xC0) == 0x80)
```

Comment:
There are *a lot* of hardcoded values. Would it be possible to give those values a name ? This could be achieved throught `#define`, or even better, with typed const values.

Here, someone new to unicode (and this code) could not grasp what this condition even test.

Actually you may even want to put those constant in their own separate header ?

Reply:
Thanks. Fixed in https://github.com/Rust-GCC/gccrs/pull/2530